Fix startup type error in initializeCaches

asalkeld opened the pull request:
The following error was getting logged:
PersistentVolumeController can't initialize caches, expected list of volumes, got:
&{TypeMeta:{Kind: APIVersion:} ListMeta:{SelfLink:/api/v1/persistentvolumes ResourceVersion:11} Items:[]}

The tests make extensive use of NewFakeControllerSource which uses api.List
instead of api.PersistentVolumeList. So use reflect to help iterate over the
items then assert the item type.

fixes #27757

jsafrane commented:
@k8s-bot ok to test

jsafrane commented on pkg/controller/persistentvolume/controller_base.go:
```diff
@@ -139,14 +140,17 @@ func (ctrl *PersistentVolumeController) initializeCaches(volumeSource, claimSour
-		return
-	}
-	for _, volume := range volumeList.Items {
+	volumeListVal := reflect.Indirect(reflect.ValueOf(volumeListObj))
```
Please add a comment why `reflect` is needed as both `List` and `PersistentVolumeList` need to be processed.

jsafrane commented:
@k8s-bot e2e test this issue: #IGNORE

jsafrane commented:
LGTM

jsafrane commented:
@k8s-bot e2e test this issue: #IGNORE

jsafrane commented:
abeling into 1.3, this basically fixes controller initialization after crash, without this patch the controller may release (and recycle!) bound volumes!

jsafrane commented:
@asalkeld, please squash your commits into one and I'll merge it.

jsafrane commented:
Ok, so @asalkeld is probably sleeping by now. @saad-ali, can you please coordinate with him to either squash the patches or merge them as they are? I'd really like to have them in Kube 1.3. This patch won't break volume controller more than it already is.

Nobody noticed this error earlier, as the volume sources behaves differently in unit tests and we don't really test controller startup with existing PVs/PVCs anywhere else. I'll write an integration test for this.

idvoretskyi commented:
@jsafrane we have enough time (~7-8 hours) to wait for @asalkeld squashing the commits and we'll be ready to merge this item (I'm against of merging as-is).

jsafrane commented:
I'm testing the patch properly and it seems it does not work as expected. Problem is that `volumeSource.List` in unit tests returns `api.List` with `*api.PersistentVolume` items, while real controller (and informer below) returns `api.PersistentVolumeList` with `api.PersistentVolume` as items, i.e. the items are not pointers.

`volume, ok := volumeListItems.Index(i).Interface().(*api.PersistentVolume)` will then fail.

idvoretskyi commented:
@jsafrane any suggestion how to fix that or it would be better to wait @asalkeld for that?

matchstick commented:
@saad-ali @jsafrane @idvoretskyi @jingxu97 @erictune  In the burn down meeting we thought it might be a good idea to have someone else look at this also. So @jingxu97 is ramping up so we can have more eyes on the issue and the fix.

jingxu97 commented:
Hi Jan,

I am trying to reproduce this issue, could you please help list the steps I
should do? Thanks a lot!


On Fri, Jun 24, 2016 at 8:44 AM, Jan Šafránek <notifications@github.com>
wrote:

> I'm testing the patch properly and it seems it does not work as expected.
> Problem is that volumeSource.List in unit tests returns api.List with
> *api.PersistentVolume items, while real controller (and informer below)
> returns api.PersistentVolumeList with api.PersistentVolume as items, i.e.
> the items are not pointers.
>
> volume, ok := volumeListItems.Index(i).Interface().(*api.PersistentVolume)
> will then fail.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28002#issuecomment-228380831>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ASSNxbG-61dfke14qO4Ob4h7mc1KnBKDks5qO_tPgaJpZM4I9cy0>
> .
>



-- 
- Jing

eparis commented:
@jingxu97 jan is in central europe. At this hour on Friday I expect him to be away from his machine...

jsafrane commented:
@jingxu97, you need to store a pv and pvc bound together in etcd and after that start the controller. With current controller it may happen that the pv is synced first without pvc in the controller cache and it will therefore be considered as released (i.e.  its phase will get Released)

jsafrane commented:
And yes, it's late here and l will check this issue tomorrow.

jingxu97 commented:
After testing, a simple fix like the following will pass the unit test.

volumeList, ok := volumeListObj.(*api.PersistentVolumeList)

I will test with e2e like Jan suggested. Please let me know if you have any
concern about the fix. Thanks!

Jing

On Fri, Jun 24, 2016 at 1:50 PM, Jan Šafránek <notifications@github.com>
wrote:

> And yes, it's late here and l will check this issue tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28002#issuecomment-228458428>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ASSNxcHDC6H_3VhJ7WbU6FxcbkxpwBpnks5qPEMXgaJpZM4I9cy0>
> .
>



-- 
- Jing

lavalamp commented on pkg/controller/persistentvolume/controller_base.go:
```diff
@@ -139,14 +140,19 @@ func (ctrl *PersistentVolumeController) initializeCaches(volumeSource, claimSour
-	}
-	for _, volume := range volumeList.Items {
+	// reflect is used here as api.List is used by the tests and api.PersistentVolumeList
+	// is passed in at startup.
```
This is extremely concerning. We need to change the tests to return the same thing the API returns.

erictune commented:
@jsafrane 

Since the PVCs and PVs are watched using separate watch operations, it seems possible that one watch could fall behind the other by an arbitrary amount of time, due to delays in various parts of the system.  So, isn't it possible for the controller to get in this same kind of state, even after syncing?

Commit:
Fix startup type error in initializeCaches

The following error was getting logged:
PersistentVolumeController can't initialize caches, expected list of volumes, got:
&{TypeMeta:{Kind: APIVersion:} ListMeta:{SelfLink:/api/v1/persistentvolumes ResourceVersion:11} Items:[]}

erictune commented:
Would it be a good idea for the PVC controller, when it sees a dangling reference, to do a cache-bypassing read to confirm the dangling reference, before recycling?

lavalamp commented:
The current version of this LGTM

lavalamp commented:
> Would it be a good idea for the PVC controller, when it sees a dangling reference, to do a cache-bypassing read to confirm the dangling reference, before recycling?

Possibly. Need to understand the locking mechanism to say for sure.

jsafrane commented:
> Since the PVCs and PVs are watched using separate watch operations, it seems possible that one watch could fall behind the other by an arbitrary amount of time, due to delays in various parts of the system. So, isn't it possible for the controller to get in this same kind of state, even after syncing?

It's not likely. Once the controller has complete view of the world (i.e. after initial startup), it can receive events from PVs and PVCs at different pace as long as the controller is the only one who binds volumes. Bad things can happen when a bound PV appears and the controller does not see appropriate counterpart PVC. The PV may get recycled or deleted, which is exactly what happened in this bug.

@asalkeld, @lavalamp thanks for your patches and reviews!

erictune commented:
If events from PVs and PVCs arrive at very different paces, doesn't that mean a bound PV could show up before the create for a PVC ever shows up?

lavalamp commented:
@erictune no, because the controller must have observed the PVC in order to make the binding in the first place. (Assumption: there's *only one* controller in operation, and no other actor can cause a binding to occur.)

@jsafrane I recall talking locking methods to make all of these operations safe even with multiple controllers with a group of red hatters and googlers a few months ago. I take it that has not been implemented yet?

jsafrane commented:
@lavalamp, even with locking as suggested on the meeting we were able to find races when two controllers were trying to bind two PVs/PVCs at the same time.
